src/main/java/gregtech/core/unification/material/internal/MaterialRegistryManager.java

```diff
@@ -0,0 +1,144 @@
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+public class MaterialRegistryManager implements IMaterialRegistryManager {
+
+    private static MaterialRegistryManager INSTANCE;
+
+    private final Object2ObjectMap<String, MaterialRegistryImpl> registries = new Object2ObjectOpenHashMap<>();
+    private final Int2ObjectMap<MaterialRegistryImpl> networkIds = new Int2ObjectOpenHashMap<>();
+
+    @Nullable
+    private Collection<Material> registeredMaterials;
+
+    private final MaterialRegistryImpl gregtechRegistry = createInternalRegistry();
+
+    private Phase registrationPhase = Phase.PRE;
+
+    private MaterialRegistryManager() {/**/}
+
+    public static MaterialRegistryManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new MaterialRegistryManager();
+        }
+        return INSTANCE;
+    }
+
+    @Nonnull
+    @Override
+    public MaterialRegistry createRegistry(@Nonnull String modid) {
+        if (getPhase() != Phase.PRE) {
+            throw new IllegalStateException("Cannot create registries in phase " + getPhase());
+        }
+
+        Preconditions.checkArgument(!registries.containsKey(modid),
+                "Material registry already exists for modid %s", modid);
+        MaterialRegistryImpl registry = new MaterialRegistryImpl(modid);
+        registries.put(modid, registry);
+        networkIds.put(registry.getNetworkId(), registry);
+        return registry;
+    }
+
+    @Nonnull
+    @Override
+    public MaterialRegistry getRegistry(@Nonnull String modid) {
+        MaterialRegistry registry = registries.get(modid);
+        return registry != null ? registry : gregtechRegistry;
+    }
+
+    @Nonnull
+    @Override
+    public MaterialRegistry getRegistry(int networkId) {
+        MaterialRegistry registry = networkIds.get(networkId);
+        return registry != null ? registry : gregtechRegistry;
+    }
+
+    @Nonnull
+    @Override
+    public Collection<MaterialRegistry> getRegistries() {
+        if (getPhase() == Phase.PRE) {
+            throw new IllegalStateException("Cannot get all material registries during phase " + getPhase());
+        }
+        return Collections.unmodifiableCollection(registries.values());
+    }
+
+    @Nonnull
+    @Override
+    public Collection<Material> getRegisteredMaterials() {
+        if (registeredMaterials == null ||
+                (getPhase() != Phase.CLOSED && getPhase() != Phase.FROZEN)) {
+            throw new IllegalStateException("Cannot retrieve all materials before registration");
+        }
+        return registeredMaterials;
+    }
+
+    @Override
+    public Material getMaterial(String name) {
```

Comment:
Why pass in a string like `modid:name` when this is what ResourceLocations were made for? This whole bit could be simplified with the use of a RL.